CDAP-15524. HTTP Batch Source. Pagination. Http client implementation

aonischuk opened the pull request:
1. Implemented all pagination types. Including custom pagination (via python code)
2. Implemented http client wrapper. Which includes proxy, basic auth, oauth2 request, TSL/SSL functionality, timeout handling.

albertshau commented:
should use mocks for the unit tests, otherwise they can break due to external changes. See https://github.com/cdapio/hydrator-plugins/blob/develop/spark-plugins/src/test/java/io/cdap/plugin/spark/test/SparkPluginTest.java#L119 for an example.

albertshau commented on pom.xml:
```diff
@@ -75,17 +75,22 @@
     <!-- this is here because project.basedir evaluates to null in the script build step -->
     <main.basedir>${project.basedir}</main.basedir>
+    <unxml.version>0.9</unxml.version>
+    <widgets.dir>widgets</widgets.dir>
```
this can be removed. It's a relic from when the build was using javascript instead of the cdap maven plugin

aonischuk replied:
done

albertshau commented on pom.xml:
```diff
@@ -103,31 +108,13 @@
-    <dependency>
-      <groupId>io.cdap.cdap</groupId>
-      <artifactId>cdap-data-pipeline</artifactId>
+      <artifactId>cdap-formats</artifactId>
```
I think this pulls in stuff like avro and grok. Since we're not using those, let's add exclusions for them, which will make the jar a little bit smaller.

aonischuk replied:
done

albertshau commented on pom.xml:
```diff
@@ -241,16 +228,100 @@
+    <!-- tests -->
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-metadata-spi</artifactId>
```
why is this needed?

aonischuk replied:
done

albertshau commented on pom.xml:
```diff
@@ -241,16 +228,100 @@
+    </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-common</artifactId>
```
why is this needed?

aonischuk replied:
done

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java:
```diff
@@ -0,0 +1,167 @@
+    Long connectTimeoutMillis = TimeUnit.SECONDS.toMillis(config.getConnectTimeout());
+    Long readTimeoutMillis = TimeUnit.SECONDS.toMillis(config.getReadTimeout());
+    RequestConfig.Builder requestBuilder = RequestConfig.custom();
+    requestBuilder.setSocketTimeout(readTimeoutMillis.intValue());
```
if these should be ints, config.getConnectTimeout() should also return an int. Otherwise there can be weird behavior due to overflow.

aonischuk replied:
done. However note that TimeUnit.SECONDS.toMillis still returns long. So conversion to int remains.

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java:
```diff
@@ -0,0 +1,167 @@
+    }
+  }
+
+  public CloseableHttpResponse executeHTTP(String uri) throws IOException {
```
is this meant to be called just once or can it be called more than once? Please add a javadoc to describe expected usage.

aonischuk replied:
done

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java:
```diff
@@ -0,0 +1,167 @@
+  private void initialize() {
+    try {
+      httpClient = createHttpClient(config);
+    } catch (IOException | URISyntaxException e) {
```
these are two pretty different types of errors, it's better not to wrap them in a runtime exception, but to let them propagate up.

I'm also confused about why createHttpClient() throws a URISyntaxException. Which method call is throwing it? Ideally this isn't an exception we would have to deal with. It would be better if this class gets passed in URI or URL objects instead of dealing with strings, with validation of those done outside this class. 

In general, validation of user input should happen as early as possible, with classes later in the logic using classes that are guaranteed to be valid.

aonischuk replied:
URISyntaxException is thown by OAuthUtil.getAccessTokenByRefreshToken() Which forms URI for retrieving accessToken by RefreshToken.

aonischuk replied:
done

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/http/SSLConnectionSocketFactoryCreator.java:
```diff
@@ -0,0 +1,130 @@
+
+    // we have to manually fall back to default keystore. SSLContext won't provide such a functionality.
+    if (keystore == null) {
+      String keyStore = System.getProperty("javax.net.ssl.keyStore");
```
How do these properties get set?

aonischuk replied:
They are unset by default. But to change default keyStore user can set them via "java -D....".

This code is there just not to break default ssl settings behavior in java when javax.net.ssl.* properties are set.

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:
```diff
@@ -15,15 +15,182 @@
+        page = getNextPage();
+      }
+
+      return (page != null && page.hasNext());
```
nit: can remove the outer parentheses

albertshau replied:
this will return a false negative if it's possible to have an empty Page. To protect against this, it's better to loop above until we get a page with entries.

aonischuk replied:
I don't think Page instance can  be null. See PageFactory#createInstance it cannot return null. Null can only be returned from getNextPage() as a sign that there are no more pages to load.

aonischuk replied:
done (removed the outer parentheses)

albertshau replied:
Empty page meaning one where page.hasNext() is false.

If page.hasNext() is false and getNextPage() would have returned a non-null, then this will terminate early.

aonischuk replied:
Empty page (with no elements) in pagination APIs usually means that you should stop paginating otherwise you'll get into endless loop.

Example API:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=80

You can go to page 100 200 500 1000 which are empty and it's endless. All with no elements.
Page 10.000:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=10000

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:
```diff
@@ -15,15 +15,182 @@
+        .timeout(config.getMaxRetryDuration(), TimeUnit.SECONDS)
+        .until(this::visitPageAndCheckStatusCode);
+    } catch (ConditionTimeoutException ex) {
+      // Retries failed. We don't need to do anything here. This will ba handled using httpStatusCode below.
```
typo: ba -> be

aonischuk replied:
done

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:
```diff
@@ -15,15 +15,182 @@
+    return !errorHandlingStrategy.shouldRetry();
+  }
+
+  protected BasePage getNextPage() throws IOException, InterruptedException {
```
annotate as nullabe

aonischuk replied:
done

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:
```diff
@@ -15,15 +15,182 @@
+
+    LOG.debug("Fetching '{}'", nextPageUrl);
+
+    PollInterval pollInterval;
```
it looks like this could be a private final that is set once in the constructor.

aonischuk replied:
done

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:
```diff
@@ -15,15 +15,182 @@
+    }
+
+    HttpEntity reponseEntity = response.getEntity();
+    byte[] responseBytes = EntityUtils.toByteArray(reponseEntity);
```
we should not be loading the entire contents into memory unless it's an error. Under normal circumstances the contents should be read as an InputStream.

aonischuk replied:
At some point we will need to read whole byte array into memory. Since it needs to be written to structured record "body" field (field type is "bytes"). Am I missing something?

albertshau replied:
That's only for the 'blob' format. Everything else should not. You can ignore this now for now, since format has been moved out.

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:
```diff
@@ -15,15 +15,182 @@
+
+    HttpEntity reponseEntity = response.getEntity();
+    byte[] responseBytes = EntityUtils.toByteArray(reponseEntity);
+    if (config.getFormat().equals(PageFormat.BLOB)) {
```
Format should only apply on success. Most of the time, an error response is not formatted the same way a successful response is formatted.

aonischuk replied:
done

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/pagination/JythonPythonExecutor.java:
```diff
@@ -0,0 +1,96 @@
+          "Unknown syntax error occurred while interpreting python code. Please make sure syntax is correct.", e);
+      }
+      throw new RuntimeException(
+        String.format("Syntax error occurred while interpreting python code: '%s'" +
```
Should not surround the exception message with text, as it makes it hard to distinguish between what comes from the exception and what comes from this line. Better to just have:

Syntax error occurred while interpreting python code: '%s'

aonischuk replied:
done

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/pagination/LinkInResponseBodyPagionationIterator.java:
```diff
@@ -0,0 +1,66 @@
+ * Every page contains a next page url. This pagination type is only supported for JSON and XML formats.
+ * Pagination happens until no next page field is present or until page contains no elements.
+ */
+public class LinkInResponseBodyPagionationIterator extends BaseHttpPaginationIterator {
```
typo: Pagio -> Pagi

aonischuk replied:
done

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/pagination/NonePagionationIterator.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Returns a single page of data
+ */
+public class NonePagionationIterator extends BaseHttpPaginationIterator {
```
typo: Pagio -> Pagi

aonischuk replied:
done

albertshau reviewed (commented):
Can you move the format stuff to a separate PR? 2000 lines is a lot and there aren't any unit tests yet. Pagination and format shouldn't share much code.

albertshau commented:
After taking a look, you also shouldn't need to setup an actual http server for the unit tests. The pagination classes can all be tested using a mock HttpClient that returns a response that you set in the test.

aonischuk commented:
1. Removed the format stuff from PR.
2. Regarding UT: I have added unit tests
3. Thank you for reviewing so much code lately.

@albertshau I have addressed your comments. Can we please merge if everything looks good?

albertshau commented:
One remaining comment, please fix and squash, then i'll merge.

aonischuk commented:
@albertshau squashed the commits. Please feel free to merge.

Ps. also added pagination unit tests.

albertshau commented on src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:
```diff
@@ -15,15 +15,194 @@
+        page = getNextPage();
+      }
+
+      return page != null && page.hasNext();
```
comment here is still not addressed.

aonischuk replied:
Empty page (with no elements) in pagination APIs usually means that you should stop paginating otherwise you'll get into endless loop.

Example API:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=80

You can go to page 100 200 500 1000 which are empty and it's endless. All with no elements.
Page 10.000:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=10000

aonischuk replied:
that's why I don't think we need any change here. Please correct me if I am wrong.

albertshau replied:
I see, makes sense

albertshau commented on src/test/java/io/cdap/plugin/http/source/common/pagination/PaginationIteratorTest.java:
```diff
@@ -0,0 +1,288 @@
+package io.cdap.plugin.http.source.common.pagination;
```
missing copyright

aonischuk replied:
done

aonischuk commented:
1. Answered your remaining comment
2. Added pagination unit tests.
3. Squashed the commit.

@albertshau please feel free to merge.

albertshau commented:
Did you leave an answer about my comment? Didn't see it.

aonischuk committed:
CDAP-15524. HTTP Batch Source. Pagination. Http client implementation

aonischuk commented:
1. Added copyright
2. Answered the comment regarding empty page getNext().

@albertshau please re-check/merge

aonischuk commented:
Thanks!
